core/runtime/jsruntime.rs

```diff
@@ -853,134 +816,52 @@ impl JsRuntime {
-            op_ctx.decl.clone(),
-            op_ctx.state.clone(),
-            op_ctx.runtime_state.clone(),
-            op_ctx.get_error_class_fn,
-            op_ctx.metrics_fn.clone(),
-          )
-        })
-        .collect();
-      context_state.borrow_mut().op_ctxs = op_ctxs;
-      context_state.borrow_mut().isolate = Some(self.v8_isolate() as _);
-
-      let raw_ptr = self.v8_isolate() as *mut v8::OwnedIsolate;
-      // SAFETY: Having the scope tied to self's lifetime makes it impossible to
-      // reference JsRuntimeState::op_ctxs while the scope is alive. Here we
-      // turn it into an unbound lifetime, which is sound because 1. it only
-      // lives until the end of this block, and 2. the HandleScope only has
-      // access to the isolate, and nothing else we're accessing from self does.
-      let isolate = unsafe { raw_ptr.as_mut() }.unwrap();
-      let scope = &mut v8::HandleScope::new(isolate);
-
-      let context = create_context(
-        scope,
-        &self.global_template_middlewares,
-        &self.global_object_middlewares,
-      );
-
-      let scope = &mut v8::ContextScope::new(scope, context);
-
-      let context = bindings::initialize_context(
-        scope,
-        context,
-        &context_state.borrow().op_ctxs,
-        self.init_mode,
-      );
-      context.set_slot(scope, context_state.clone());
-
-      let loader = options
-        .module_loader
-        .unwrap_or_else(|| Rc::new(NoopModuleLoader));
-      let module_map = Rc::new(ModuleMap::new(loader));
-      if self.init_mode.is_from_snapshot() {
-        let snapshotted_data =
-          snapshot_util::get_snapshotted_data(scope, context);
-        module_map.update_with_snapshotted_data(scope, snapshotted_data);
-      }
-      context.set_slot(scope, module_map.clone());
-
-      let realm = JsRealmInner::new(
-        context_state,
-        v8::Global::new(scope, context),
-        module_map,
-        self.inner.state.clone(),
-        false,
-      );
-
-      let mut state = self.inner.state.borrow_mut();
-      // Poison the fast path for realms
-      self.additional_realms = true;
-      state.known_realms.push(realm.clone());
-      JsRealm::new(realm)
-    };
-
-    self.init_extension_js(&realm)?;
-
-    // If the embedder has requested to clear the module map resulting from
-    // extensions, possibly with exceptions.
-    if let Some(preserve_snapshotted_modules) =
-      self.preserve_snapshotted_modules
-    {
-      realm
-        .0
-        .module_map()
-        .clear_module_map(preserve_snapshotted_modules);
-    }
-
-    Ok(realm)
-  }
-
   #[inline]
   pub fn handle_scope(&mut self) -> v8::HandleScope {
-    self.main_realm().handle_scope(self.v8_isolate())
+    let isolate = &mut self.inner.v8_isolate;
+    self.inner.main_realm.handle_scope(isolate)
+  }
+
+  #[inline(always)]
+  /// Create a scope on the stack with the given context
+  fn with_runtime_scope<'s, T>(
```

Comment:
Shouldn't this be called `with_context_scope` since you are creating `v8::ContextScope`?

Reply:
Good call, I'll rename it.